src/host-config.ts

```diff
@@ -0,0 +1,182 @@
+
+const memoizedAuthRegex = _.memoize(
+	(proxyField: string) => new RegExp(proxyField + '\\s*=\\s*"(.*)"\\s*;'),
+);
+
+const memoizedRegex = _.memoize(
+	proxyField => new RegExp(proxyField + '\\s*=\\s*([^;\\s]*)\\s*;'),
+);
+
+async function readProxy(): Promise<ProxyConfig | undefined> {
+	const conf: ProxyConfig = {};
+	let redsocksConf: string;
+	try {
+		redsocksConf = await fs.readFile(redsocksConfPath, 'utf-8');
+	} catch (e) {
+		if (!ENOENT(e)) {
+			throw e;
+		}
+		return;
+	}
+	const lines = redsocksConf.split('\n');
+
+	for (const line of lines) {
+		for (const proxyField of proxyFields) {
+			let match: string[] | null = null;
+			if (isAuthField(proxyField)) {
+				match = line.match(memoizedAuthRegex(proxyField));
+			} else {
+				match = line.match(memoizedRegex(proxyField));
+			}
+
+			if (match != null) {
+				conf[proxyField] = match[1];
+			}
+		}
+	}
+
+	try {
+		const noProxy = await fs.readFile(noProxyPath, 'utf-8');
+		conf.noProxy = noProxy.split('\n');
+	} catch (e) {
+		if (!ENOENT(e)) {
+			throw e;
+		}
+	}
+	return conf;
+}
+
+function generateRedsocksConfEntries(conf: ProxyConfig): string {
+	let val = '';
+	for (const field of proxyFields) {
+		let v = conf[field];
+		if (v != null) {
+			if (isAuthField(field)) {
+				v = `"${v}"`;
+			}
+			val += `\t${field} = ${v};\n`;
+		}
+	}
+	return val;
+}
+
+async function setProxy(maybeConf: ProxyConfig | null): Promise<void> {
+	if (_.isEmpty(maybeConf)) {
+		try {
+			await Promise.all([fs.unlink(redsocksConfPath), fs.unlink(noProxyPath)]);
+		} catch (e) {
+			if (!ENOENT(e)) {
+				throw e;
+			}
+		}
+	} else {
+		// We know that maybeConf is not null due to the _.isEmpty check above,
+		// but the compiler doesn't
+		const conf = maybeConf as ProxyConfig;
+		await mkdirp(proxyBasePath);
+		if (_.isArray(conf.noProxy)) {
+			await writeFileAtomic(noProxyPath, conf.noProxy.join('\n'));
+		}
+		const redsocksConf = `${redsocksHeader}${generateRedsocksConfEntries(
+			conf,
+		)}${redsocksFooter}`;
+		await writeFileAtomic(redsocksConfPath, redsocksConf);
+	}
+
+	await systemd.restartService('resin-proxy-config');
+	await systemd.restartService('redsocks');
+}
+
+const hostnamePath = path.join(constants.rootMountPoint, '/etc/hostname');
+async function readHostname() {
+	const hostnameData = await fs.readFile(hostnamePath, 'utf-8');
+	return _.trim(hostnameData);
+}
+
+async function setHostname(val: string, configModel: Config) {
+	await configModel.set({ hostname: val });
+	await systemd.restartService('resin-hostname');
+}
+
+// Don't use async/await here to maintain the bluebird
+// promises being returned
+export function get(): Bluebird<HostConfig> {
+	return Bluebird.join(readProxy(), readHostname(), (proxy, hostname) => {
+		return {
+			network: {
+				proxy,
+				hostname,
+			},
+		};
+	});
+}
+
+export function patch(conf: HostConfig, configModel: Config): Bluebird<void> {
+	const promises = [];
+	if (conf != null && conf.network != null) {
```

Comment:
Based on the typings `conf` can never be null

Reply:
This is just what it was in the coffeescript, as I say, when every call site is TS I can be sure to take this out and that the types are stringent enough, but until then I'm really not comfortable taking out the checking, and I'd much prefer the types to as strict as possible.